spec/openai/client/http_spec.rb

```diff
@@ -156,22 +153,31 @@
         end
       end
 
-      context "when called with a string containing an error" do
-        let(:chunk) do
-          <<-CHUNK
-            data: { "foo": "bar" }
-            error: { "message": "A bad thing has happened!" }
-          CHUNK
-        end
+      context "when OpenAI returns an HTTP error" do
+        let(:chunk) { "{\"error\":{\"message\":\"A bad thing has happened!\"}}" }
+        let(:env) { Faraday::Env.new(status: 500) }
 
-        it "does not raise an error" do
-          expect(user_proc).to receive(:call).with(JSON.parse('{ "foo": "bar" }'))
+        it "does not raise an error and calls the user proc with the error parsed as JSON" do
           expect(user_proc).to receive(:call).with(
-            JSON.parse('{ "message": "A bad thing has happened!" }')
+            {
+              "error" => {
+                "message" => "A bad thing has happened!"
+              }
+            }
           )
 
           expect do
-            stream.call(chunk)
+            stream.call(chunk, 0, env)
+          end.not_to raise_error
+        end
+      end
+
+      context "when called with JSON split across chunks" do
```

Comment:
This is the meat of this PR. ruby-openai now becomes resilient to packet fragmentation at awkward locations.

Reply:
Hi @atesgoral, I'm an engineer at OpenAI investigating this response. Do you have any logs (ideally from `curl` showing the full response stream) to demonstrate this issue? Please feel free to email me at atty@openai.com. Thanks!

Reply:
@athyuttamre I haven't recently seen this happen when directly talking to the OpenAI client, but I recall seeing it a while back, when network conditions were bad. But it can still happen when a proxy or some other network element buffers and chops up the text/event-stream chunks at non-event-stream boundaries.

Reply:
I don't think this is an OpenAI problem at all. Clients just need to be resilient and do buffered/proper parsing of event streams.

Reply:
For posterity, here are two chunks (stream fragments) from an earlier reproduction of the parser issue that gets unearthed by buffering + rechunking:

Chunk 1:

```
data: {"id":"chatcmpl-83QVo11UROyI8HUeixAo25Vjx0SA5","object":"chat.completion.chunk","created":1695827588,"model":"gpt-4-0613","choices":[{"index":0,"delta":{"role":"assistant","content":""},"finish_reason":null}]}
data: {"id":"chatcmpl-83QVo11UROyI8HUeixAo25Vjx0SA5","object":"chat.completion.chunk","created":1695827588,"model":"gpt-4-0613","choices":[{"index":0,"delta"
```

Chunk 2:

```
:{"content":"Hello"},"finish_reason":null}]}
```